brainglobe_stitch/image_mosaic.py

```diff
@@ -340,3 +346,366 @@ def read_big_stitcher_transforms(self) -> None:
+            The depth of the resolution pyramid.
+        downscale_factors: Tuple[int, int, int], default: (1, 2, 2)
+            The factors to downscale the image by in the z, y, x dimensions.
+        """
+        z_size, y_size, x_size = self.tiles[0].data_pyramid[0].shape
+        output_file = h5py.File(output_path, mode="w")
+
+        # Metadata is in x, y, z order for Big Data Viewer
+        downscale_factors_array = np.array(downscale_factors)
+        resolutions = np.ones((pyramid_depth, 3), dtype=np.int16)
+
+        for i in range(1, pyramid_depth):
+            resolutions[i, :] = downscale_factors_array[-1::-1] ** i
+
+        channel_ds_list: List[List[h5py.Dataset]] = []
+        for i in range(self.num_channels):
+            chunk_shape: Tuple[int, ...] = (128, 128, 128)
+            # Write the resolutions and subdivisions for each channel
+            output_file.require_dataset(
+                f"s{i:02}/resolutions",
+                data=resolutions,
+                dtype="i2",
+                shape=resolutions.shape,
+            )
+            output_file.create_dataset(
+                f"s{i:02}/subdivisions",
+                dtype="i2",
+                shape=resolutions.shape,
+            )
+
+            if np.any(np.array(fused_image_shape) < chunk_shape):
+                chunk_shape = fused_image_shape
+
+            # Create the datasets for each resolution level
+            ds_list: List[h5py.Dataset] = []
+            ds = output_file.require_dataset(
+                f"t00000/s{i:02}/0/cells",
+                shape=fused_image_shape,
+                chunks=chunk_shape,
+                dtype="i2",
+            )
+            ds_list.append(ds)
+            # Set the chunk shape for the first resolution level
+            # Flip the shape to match the x, y, z order
+            output_file[f"s{i:02}/subdivisions"][0] = chunk_shape[::-1]
+            new_shape = (
+                (fused_image_shape[0] + (downscale_factors[0] > 1))
+                // downscale_factors[0],
+                (fused_image_shape[1] + (downscale_factors[1] > 1))
+                // downscale_factors[1],
+                (fused_image_shape[2] + (downscale_factors[2] > 1))
+                // downscale_factors[2],
+            )
+
+            for j in range(1, len(resolutions)):
+                if np.any(np.array(new_shape) < chunk_shape):
+                    chunk_shape = new_shape
+
+                down_ds = output_file.require_dataset(
+                    f"t00000/s{i:02}/{j}/cells",
+                    shape=new_shape,
+                    chunks=chunk_shape,
+                    dtype="i2",
+                )
+
+                ds_list.append(down_ds)
+                # Set the chunk shape for the other resolution levels
+                # Flip the shape to match the x, y, z order
+                output_file[f"s{i:02}/subdivisions"][j] = chunk_shape[::-1]
+
+                new_shape = (
```

Comment:
Could this block of code be factored into a function to avoid duplication with above? It might also make this section easier to understand, maybe a quick comment `# update new shape before the next downscaling iteration` just as confused me a bit at first

Reply:
Factored it out and simplified this section.